controllers/resource_controller.go

```diff
@@ -120,11 +152,35 @@ func (r *ResourceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 
 	log.Info("found resource", "resource", obj)
 
-	*result, retErr = r.reconcile(ctx, obj)
-	return *result, retErr
+	result, retErr = r.reconcile(ctx, componentVersion, obj)
+	return result, retErr
+}
+
+func (r *ResourceReconciler) shouldReconcile(ctx context.Context, cv *v1alpha1.ComponentVersion, obj *v1alpha1.Resource) (bool, error) {
+	// If there is a mismatch between the observed generation of a component version, we trigger
+	// a reconcile. There is either a new version available or a dependent component version
+	// finished its reconcile process.
+	if obj.Status.LastObservedComponentVersionGeneration != cv.Status.ObservedGeneration {
```

Comment:
rather than tracking observed generation for component versions I think it would be better to track the ReconciledVersion field of the ComponentVersion. This is because the ComponentVersion spec supports version constraints which means that new versions can be reconciled without a change in spec and the accompanying change in generation.